port/linux/omrvmem.c

```diff
@@ -811,7 +811,15 @@ omrvmem_reserve_memory_ex(struct OMRPortLibrary *portLibrary, struct J9PortVmemI
 
 		/* Make sure that the alignment is a multiple of both requested alignment and page size (enforces that arguments are powers of two and, thus, their max is their lowest common multiple) */
 		if ((0 == largePageMinimumGranule) || (0 == (largePageAlignmentInBytes % largePageMinimumGranule))) {
-			memoryPointer = reserveLargePages(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, params->pageSize, largePageAlignmentInBytes, params->options, params->mode);
+#if __GLIBC_PREREQ(2,27)
+			if (0 != (OMRPORT_VMEM_MEMORY_MODE_USE_MEMFD_CREATE & params->mode)) {
+				uintptr_t mode = params->mode | OMRPORT_VMEM_MEMORY_MODE_HUGE_PAGES;
+				memoryPointer = getMemoryInRangeForDefaultPages(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, largePageAlignmentInBytes, params->options, mode);
+			} else
+#endif /* __GLIBC_PREREQ(2,27) */
+			{
+				memoryPointer = reserveLargePages(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, params->pageSize, largePageAlignmentInBytes, params->options, params->mode);
+			}
```

Comment:
I somehow don't like making getMemoryInRangeForDefaultPages (and default_pageSize_reserve_memory) deal with large pages when they were originally meant for _default_ page size.

I can spot at least one potential problem: getMemoryInRangeForDefaultPages  ends up calling default_pageSize_reserve_memory with page size of  hardcoded PPG_vmem_pageSize[0], what might be the value eventually stored in VmemIdentifier, even though it was attempted to use with large pages. 

I wonder if it was possible to keep this intact and make separation inside reserveLargePages (for example, create reserveLargePagesWithMemfd and getMemoryInRangeForLargePagesWithMemfd  where you can deviate to). 

I understand the motivation (and valid so) is to reuse mmap-ing code later that for now is used only for default pages. perhaps some of that logic can be isolated under more generic method (neither specific for default nor large pages)

Reply:
Yeah I'm more incline to use a more generic method that neither specify default nor large pages. Also agree that using `getMemoryInRangeForDefaultPages` for large pages is not desirable even though this works because the decision to map virtual memory with large pages is done before we call `omrvmem_reserve_memory_ex`. But again, moving the logic into `reserveLargePages` like Aleks suggested would make more sense.

Reply:
Looks like there is change of behaviour for Large Pages - old code did try to allocate in Small Pages by calling `getMemoryInRangeForDefaultPages()`. And this important path to get memory in Small Pages because LP might be not available run time. Did I miss something and this path is handled somehow?

Reply:
Yes, I talked to Aleks and we were waiting on some reviews. My thinking was since we were enabling double mapping with large pages I added logic to reserve large pages with mmap instead of shmget and shmat when possible. But we thought it would be better to remove this update from this PR and include it in a separate one. What you think?

Reply:
We should preserve old behaviour (but not code necessary) at least while we are using old API.
For example there is base scenario:
- xLinux
- 1m pages are provisioned (available)
- GC: requests Object Heap in 1m pages by default
- Port: An allocation in 1m pages can not be done due lack of them
- Port: An attempt to get memory in 4k pages instead should be issued in the same request, and... success
- GC: request is fulfilled in 4k pages

I understand that with new API the structure of calls should be different. And it might be a trigger to do change of logic suitable for both worlds: new and old. But functionality from old code must be covered (or explicitly rejected as unnecessary - this is different subject)

Reply:
Totally agree. I think the confusion here is that should we allow double map with huge pages or not? If we answer yes, then we would need to create the heap using `mmap` with huge pages for that to work; hence my logic to include the extra change to reserve memory with `mmap` with large pages. All those scenarios listed is taken care of, the major change would be to instead using `shmget` + `shmat` would be to use `mmap`. In my humble opinion it would be logical to include both changes related to huge pages in this PR, or don't deal with huge pages here and create another PR that deals with that. What you think?

Reply:
If we look to `omrvmem_reserve_memory_ex()` for first look logic is based on page size. But in real this organization based on memory allocation API type (`shmget()` vs `mmap()`). And this might be good base for our needs too.
What if we use this very general logic:
```
 /* --- input parameters correctness check - only once - return immediately if any problem --- */

    memoryPointer = NULL;

#if defined(MFD_ALLOCATION_AVAILABLE) /* some name of preprocessor flag */
    if (mfdAllocationEnabled) { /* mfd might be available but not allowed - we will need to add feature key for this(?) */

        /* an attempt to allocate ANY page size using memfd_create */
        if (thisIsLargePage) {
            /* add required flags if necessary*/
        }
        memoryPointer = reserveMemoryUsingMfd(...);

    } /* else */  /* we can put else here if we want to be mutually exclusive and do not try mmap/shmget if mfd failed  */
#endif /* defined(MFD_ALLOCATION_AVAILABLE) */
    {
        if ((NULL == memoryPointer) && largePageRequested) {

            /* we can try shm* type of allocation */
            memoryPointer = reserveMemoryUsingShm(...);  /* it is probably existing reserveLargePages() ?? */
        }

        if (NULL == memoryPointer) {

            /* allocation using mmap() */
            /* set page size small if necessary */
            memoryPointer = reserveMemoryUsingMmap(...);
        }
    }

    if (NULL == memoryPointer) {
        /* no success - error messages */
    } else {
        /* post actions - set NUMA etc */
    }
```
There is no double mapping logic in this picture (yet) but can it work?  What you think?

Reply:
This could work for sure. And we could even further simplify by combining `reserveMemoryUsingMfd` and `reserveMemoryUsingMmap`, because they take the same path, meaning both end up using `mmap()` to reserve the heap. where the difference would be:
- `reserveMemoryUsingMfd`: creates heap with file descriptor.
- `reserveMemoryUsingMmap`: creates anonymous heap.
Also, in order for `reserveMemoryUsingMfd` to work with large pages we would need to `mmap()` the heap using large pages, more specifically, use the `MAP_HUGETLB` flag. My concern is that, up until now we've been using (`shmget()` vs `mmap()`) to differentiate between large and default page size paths respectively. Now that we're introducing `memfd_create()` with huge pages we're breaking this pattern; where whenever the system supports large pages, as well as `memfd_create()`, we'll take the `mmap()` path instead of the `shmget()` path. Hence, we could leverage this by also trying to reserve large pages through `mmap()` instead of `shmget()` whenever we can, or would that be overkill?

Reply:
I have a suspicion that we are going to be requested to support LP allocation using `shmget()` by default to preserve existing behaviour. However we should have flexibility to change this default mode runtime. Or even make new behaviour default